VDA-XP-Reward/contracts/VDAXPReward.sol

```diff
@@ -0,0 +1,277 @@
+    error EmptyClaimData();
+    error InvalidXP(address didAddress, string typeId, uint xp);
+    error InvalidProofTime(address didAddress, string typeId, uint16 year, uint8 month);
+    error InvalidProof(address didAddress, string typeId, uint xp, bytes signature);
+    error InsufficientTokenAmount(uint requestedAmount, uint currentAmount);
+    error DuplicatedRequest(address didAddress, string typeId, uint xp, bytes signature);
+    error DuplicatedUniqueId(address didAddress, string typeId, string uniqueId, uint xp);
+
+    function __VDAXPReward_init(IERC20Upgradeable token, IVeridaDIDRegistry didRegistry) public initializer {
+        __VDAVerificationContract_init();
+        __VDAXPReward_init_unchained(token, didRegistry);
+    }
+
+    function __VDAXPReward_init_unchained(IERC20Upgradeable token, IVeridaDIDRegistry didRegistry) internal {
+        rewardToken = token;
+        didRegistryContract = didRegistry;
+        rateDenominator = 10000000; // Set up rate from 0.000001
+    }
+
+    /**
+     * @dev See {IVDAXPReward}
+     */
+    function getTokenAddress() external virtual view override returns(address) {
+        return address(rewardToken);
+    }
+
+    /**
+     * @dev See {IVDAXPReward}
+     */
+    function getDIDRegistryAddress() external virtual view override returns(address) {
+        return address(didRegistryContract);
+    }
+
+    /**
+     * @dev See {IVDAXPReward}
+     */
+    function getRateDenominator() external virtual view override returns(uint32) {
+        return rateDenominator;
+    }
+
+    /**
+     * @dev See {IVDAXPReward}
+     */
+    function setRateDenominator(uint32 denominator) external virtual override onlyOwner {
+        if (rateDenominator == denominator || denominator == 0) {
+            revert InvalidValue();
+        }
+        uint32 orgVal = rateDenominator;
+        rateDenominator = denominator;
+        emit UpdateRateDenominator(orgVal, denominator);
+    }
+
+    /**
+     * @dev See {IVDAXPReward}
+     */
+    function getConversionRate() external virtual view override returns(uint32) {
+        return conversionRate;
+    }
+
+    /**
+     * @dev See {IVDAXPReward}
+     */
+    function setConversionRate(uint32 newRate) external virtual override onlyOwner {
+        if (newRate == 0 || newRate == conversionRate) {
+            revert InvalidValue();
+        }
+        uint32 orgVal = conversionRate;
+        conversionRate = newRate;
+        emit UpdateConversionRate(orgVal, newRate);
+    }
+
+    /**
+     * @notice Validate the time of proof
+     * @dev Separated for `stack too deep`
+     * @param didAddress - DID address. Used to revert
+     * @param curTime - Current time
+     * @param info - Claim information
+     */
+    function _validateProofTime(address didAddress, DateTime._DateTime memory curTime, ClaimInfo calldata info) internal virtual view {
+        if (curTime.year != info.issueYear || curTime.month != (info.issueMonth + 1)) {
```

Comment:
`curTime.year != info.issueYear`

I think this will return a false positive when current date is Jan 2025 (issuer year will not match current year).

Reply:
Please add failed unit tests that cover this scenario.

Reply:
Already, consider all cases. It is checked by the `else if` statement.
Test case was already added:

Reply:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/224bfda1-1fa9-4c2d-b27d-5ff4d15fdf4f)

Reply:
You need to test blocktime = december, claim time is january. And claim is accepted. You will need to set the blockcchain blocktime to be December to do this properly.

Reply:
Added the test case:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/b154c90f-e62f-49e4-bc7f-d7c8768073c3)

Reply:
@ITStar10 

This new test you have added succeeds, but it should fail.

You set the blocktime to be January 2024, but `CLAIM_GAMER31` has month of April 2024:

```
issueYear: 2024,
issueMonth: 4,
```

Reply:
Block time is set as 2025/1 by the following code:
```ts
blockTime.setFullYear(blockTime.getFullYear() + 1);
blockTime.setMonth(0);
await time.increaseTo(blockTime.getTime() / 1000);
```
Here, it increases the year.

And, the issueYear and issueMonth of `CLAIM_GAMER31` is updated in the `checkClaimSuccess()` function.
```ts
const checkClaimSuccess = async (claimInfos: ClaimData[], rewardXPs: number[]) => {
                const claimData = [];
                const [issueYear, issueMonth] = await getProofIssueTime(); //Here get the time that is 1 month before of 
 current blockchain time: 2024/12
                ...

                for (let i = 0; i < rewardXPs.length; i++) {
                    ...

                    const claimInfo: ClaimData = {...claimInfos[i]};
                    claimInfo.issueYear = issueYear;  //  Here set 2024
                    claimInfo.issueMonth = issueMonth; // Here set 12
                    ...

```

The `getProofIssueTime()` function gets the time that is 1 month before of the current blockchain time:

Here attach the result of `Claim in January` test:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/6b5470a9-77f7-4c16-ab6a-10ec269b85f5)